In [1]:
import warnings
warnings.filterwarnings("ignore")
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from datetime import datetime
import seaborn as sns
%matplotlib inline
import wrangle as w
import os
import env
import outliers as o
import explore as e
import functions as f

In [2]:
df = w.prep_log_data()

In [3]:
df.head()

time                                 path  user_id  cohort_id  \
date                                                                            
2018-01-26  09:55:03                                    /        1          8   
2018-01-26  09:56:02                              java-ii        1          8   
2018-01-26  09:56:05  java-ii/object-oriented-programming        1          8   
2018-01-26  09:56:06   slides/object_oriented_programming        1          8   
2018-01-26  09:56:24            javascript-i/conditionals        2         22   

                      ip  id     name     slack start_date   end_date  \
date                                                                    
2018-01-26  97.105.19.61   8  Hampton  #hampton 2015-09-22 2016-02-06   
2018-01-26  97.105.19.61   8  Hampton  #hampton 2015-09-22 2016-02-06   
2018-01-26  97.105.19.61   8  Hampton  #hampton 2015-09-22 2016-02-06   
2018-01-26  97.105.19.61   8  Hampton  #hampton 2015-09-22 2016-02-06   
2018-01-26  97.105.19.61  22    Teddy    #teddy 2018-01-08 2018-05-17   

                    created_at          updated_at  program_id  
date                                                            
2018-01-26 2016-06-14 19:52:26 2016-06-14 19:52:26           1  
2018-01-26 2016-06-14 19:52:26 2016-06-14 19:52:26           1  
2018-01-26 2016-06-14 19:52:26 2016-06-14 19:52:26           1  
2018-01-26 2016-06-14 19:52:26 2016-06-14 19:52:26           1  
2018-01-26 2018-01-08 13:59:10 2018-01-08 13:59:10           2

##  1. Which lesson appears to attract the most traffic consistently across cohorts (per program)?

#### Program 1

In [4]:
f.print_top_lessons1(df)

Top 4 most visited lessons:
Lesson Path: /
Total Traffic: 1681
---------------------------
Lesson Path: index.html
Total Traffic: 1011
---------------------------
Lesson Path: javascript-i
Total Traffic: 736
---------------------------
Lesson Path: html-css
Total Traffic: 542
---------------------------


#### For program 1 the most visited lesson was javascript-1

#### Program 2

In [5]:
f.print_top_lessons2(df)

Top 4 most visited lessons:
Lesson Path: /
Total Traffic: 35814
---------------------------
Lesson Path: javascript-i
Total Traffic: 17457
---------------------------
Lesson Path: toc
Total Traffic: 17428
---------------------------
Lesson Path: search/search_index.json
Total Traffic: 15212
---------------------------


#### For program 2 the most visited lesson was javascript-1

#### Program 3

In [6]:
f.print_top_lessons3(df)

Top 4 most visited lessons:
Lesson Path: /
Total Traffic: 8358
---------------------------
Lesson Path: search/search_index.json
Total Traffic: 2203
---------------------------
Lesson Path: classification/overview
Total Traffic: 1785
---------------------------
Lesson Path: 1-fundamentals/modern-data-scientist.jpg
Total Traffic: 1655
---------------------------


#### For program 3 the most visited lesson was classification/overview

#### Program 4

In [7]:
f.print_top_lessons4(df)

Top 4 most visited lessons:
Lesson Path: content/html-css
Total Traffic: 2
---------------------------
Lesson Path: /
Total Traffic: 1
---------------------------
Lesson Path: content/html-css/gitbook/images/favicon.ico
Total Traffic: 1
---------------------------
Lesson Path: content/html-css/introduction.html
Total Traffic: 1
---------------------------


#### For program 4 the most visited lesson was content/html-css

## 2. Is there a cohort that referred to a lesson significantly more than other cohorts seemed to gloss over?

## 3. Are there students who, when active, hardly access the curriculum? If so, what information do you have about these students?

## 4. Is there any suspicious activity, such as users/machines/etc accessing the curriculum who shouldn’t be? Does it appear that any web-scraping is happening? Are there any suspicious IP addresses?

## 7.  Which lessons are least accessed?

In [10]:
lessons = df[["program_id", "cohort_id", "path"]]

result_dataframe = f.process_lessons_data(lessons)

In [11]:
result_dataframe

path  traffic
63           10-nlp/2-regular-expressions        1
64                       10-nlp/3-acquire        1
68                       10-nlp/exercises        1
99    11-distributed-ml/3-getting-started        1
100           11-distributed-ml/4-acquire        1
...                                   ...      ...
2203                         users/1/edit        1
2204                            users/167        1
2205                         users/2/edit        1
2206                            users/913        1
2215                        web-design/ux        1

[185 rows x 2 columns]

## Looking through the lessons with the least traffic which was 1, there were lessons that no longer exist in the current curriculum and had testing titles such as asdfasdf as the lesson name. This tells me that there may at some point have been testing website pages. I would need a current website paths to ensure I am not delivering lessons that were "least accessed" that were never technically a lesson to began with. Using the current data, here are the 5 least visited lessons

- nlp/2-regular-expressions
- nlp/3-acquire
- nlp/exercises
- distributed-ml/3-getting-started
- distributed-ml/4-acquire

*** There are 185 more lessons with a traffic of 1, this is just 5 ***